sparse-index.c

```diff
@@ -295,6 +295,7 @@ void ensure_full_index(struct index_state *istate)
 
 	/* Copy back into original index. */
 	memcpy(&istate->name_hash, &full->name_hash, sizeof(full->name_hash));
+	memcpy(&istate->dir_hash, &full->dir_hash, sizeof(full->dir_hash));
```

Comment:
Wow. Thanks for finding this missing memory copy. This applies to the version of sparse-index in upstream's `master`, so we should send it up as its own bugfix (independent of other enhancements).

Reply:
I wonder whether it would be worth running `t1091`, `t1092`, `t7519` and `t7817` through `valgrind` on Linux and macOS, which _may_ have caught this?

Reply:
I tried to set up a GitHub workflow to do this: https://github.com/dscho/git/runs/2998707560?check_suite_focus=true

What a rabbit hole. Apparently, `brew install valgrind` fails because it says it requires Linux... but then I found an unofficial fork that builds on macOS, except it fails because of a missing syscall (I [reported this](https://github.com/LouisBrunner/valgrind-macos/issues/38)).

On the Linux side, it's just really slow. I also suspect that it runs the tests over and over again because it is still running, at the time of writing it ran for over 45 minutes (granted, some 4-6 minutes of that is the _build_, not the _test_).

Reply:
Ah, so it took ~25 minutes to run the tests (in parallel, `t1092` took this long on its own), and it ran things _twice_ because `linux-gcc` runs the test suite twice, once with default options, and then with a ton of options overridden (such as testing SHA-256 instead of SHA-1). But it _did_ finish eventually.

I'm not sure what to do about testing on macOS, though.